File: mixin-base/src/main/java/net/yakclient/mixin/base/internal/methodadapter/tree/TreeBRPatternMatcher.java
package net.yakclient.mixin.base.internal.methodadapter.tree;

import net.yakclient.mixin.base.internal.bytecode.ByteCodeUtils;
import net.yakclient.mixin.base.internal.instruction.Instruction;
import org.objectweb.asm.tree.AbstractInsnNode;
import org.objectweb.asm.tree.InsnList;

/**
 * Defines the pattern for inserting mixins before every return
 * statement in the specified instructions.
 * <p>
 * Type: Before Return
 *
 * @author Durgan McBroom
 */
public class TreeBRPatternMatcher extends TreeMixinPatternMatcher {
    public TreeBRPatternMatcher(Instruction instructions) {
        super(instructions);
    }

    @Override
    public void transform(InsnList insn) {
        for (AbstractInsnNode node : insn) {
            if (ByteCodeUtils.isReturn(node.getOpcode())) this.insertInsn(node, insn);
        }
    }
}

File: base/src/main/kotlin/net/yakclient/mixins/base/internal/methodadapter/tree/TreeBRPatternMatcher.kt
package net.yakclient.mixins.base.internal.methodadapter.tree

import net.yakclient.mixins.base.internal.bytecode.ByteCodeUtils.isReturn
import net.yakclient.mixins.base.internal.instruction.Instruction
import org.objectweb.asm.tree.InsnList

/**
 * Defines the pattern for inserting mixins before every return
 * statement in the specified instructions.
 *
 *
 * Type: Before Return
 *
 * @author Durgan McBroom
 */
class TreeBRPatternMatcher(
    instructions: Instruction
) : TreeMixinPatternMatcher(instructions) {
    override fun transform(insn: InsnList) {
        for (node in insn) {
            if (isReturn(node.opcode)) this.insertInsn(node, insn)
        }
    }
}